src/python/pants/core/goals/lint.py

```diff
@@ -198,18 +228,20 @@ def rules():
     def debug_hint(self) -> str:
         return self.name
 
-    if TYPE_CHECKING:
+    @dataclass(frozen=True)
+    @runtime_ignore_subscripts
+    class SubPartition(Generic[_PartitionElementT]):
+        elements: Tuple[_PartitionElementT, ...]
+        key: Any
 
-        class SubPartition(Collection[_ElementT]):
-            pass
+    _SubPartitionBase = SubPartition
```

Comment:
These are required so we can use them as the "base" class. Otherwise we'd hit an infinite recursion trying to lookup the class.